Refactor DataprocCreateCluster operator to use simpler interface

turbaszek opened the pull request:
DataprocCreateCluster requires now:
- cluster config
- cluster name
- project id

In this way users don't have to pass project_id two times (in cluster definition and as parameter). The cluster object is built in `create_cluster` hook method

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

turbaszek committed:
Fix duration object in Dataproc cluster generator

turbaszek commented on airflow/providers/google/cloud/operators/dataproc.py:
```diff
@@ -289,26 +289,24 @@ def _build_gce_cluster_config(self, cluster_data):
             worker_type_uri = \
-                "https://www.googleapis.com/compute/v1/projects/{}/zones/{}/machineTypes/{}".format(
-                    self.project_id, self.zone, self.worker_machine_type)
+                f"projects/{self.project_id}/zones/{self.zone}/machineTypes/{self.worker_machine_type}"
```
Not related but simplifies things. Reference:
>Optional. The Compute Engine machine type used for cluster instances.
>A full URL, partial URI, or short name are valid. Examples:
> - https://www.googleapis.com/compute/v1/projects/[projectId]/zones/us-east1-a/machineTypes/n1-standard-2
> - projects/[projectId]/zones/us-east1-a/machineTypes/n1-standard-2
> - n1-standard-2

turbaszek committed:
fixup! Fix duration object in Dataproc cluster generator

olchas commented on airflow/providers/google/cloud/operators/dataproc.py:
```diff
@@ -240,10 +240,10 @@ def _get_init_action_timeout(self):
         if match:
             if match.group(2) == "s":
-                return self.init_action_timeout
+                return {"seconds": self.init_action_timeout}
```
It seems value under `seconds` key will be of `str` type in this case, and an `int` type in the other - is that intended/valid?

turbaszek replied:
Fixed

olchas commented on airflow/providers/google/cloud/operators/dataproc.py:
```diff
@@ -157,8 +157,8 @@ class ClusterGenerator:
-        project_id: Optional[str] = None,
-        cluster_name: Optional[str] = None,
+        project_id: str,
+        cluster_name: str,
```
Should a note about these arguments no longer being optional be added somewhere (UPDATING.md perhaps)?

turbaszek replied:
I added fallback for default project id. The cluster name has required any way

mik-laj commented on airflow/providers/google/cloud/operators/dataproc.py:
```diff
@@ -157,8 +157,8 @@ class ClusterGenerator:
         self,
-        project_id: Optional[str] = None,
-        cluster_name: Optional[str] = None,
+        project_id: str,
```
Should we add fallback for project ID from service account?

turbaszek replied:
Added

mik-laj commented on airflow/providers/google/cloud/operators/dataproc.py:
```diff
@@ -157,8 +157,8 @@ class ClusterGenerator:
-        project_id: Optional[str] = None,
-        cluster_name: Optional[str] = None,
+        project_id: str,
+        cluster_name: str,
```
It depends if this parameter was actually optional before. I think cluster_name was a required parameter, but we had the wrong method signature.  In that case, this change doesn't change anything, so it's not a breaking change.

turbaszek replied:
Yes, this change in fact fixes the cluster generator

turbaszek committed:
fixup! fixup! Fix duration object in Dataproc cluster generator

olchas commented on airflow/providers/google/cloud/operators/dataproc.py:
```diff
@@ -240,10 +241,10 @@ def _get_init_action_timeout(self):
         if match:
             if match.group(2) == "s":
-                return self.init_action_timeout
+                return {"seconds": int(self.init_action_timeout)}
```
```suggestion
                return {"seconds": int(match.group(1))}
```

olchas replied:
Actually, converting to float first seems to be unnecessary, as a float value will not match a regex anyway.

turbaszek replied:
Not sure if I understand

olchas replied:
You cannot convert `init_action_timeout` to `int`, as `init_action_timeout` has a character `s` at the end. You have to convert `match.group(1)` instead. I edited the original suggestion.

olchas replied:
@turbaszek it seems you have reverted to the previous version here in one of latter commits :)

turbaszek replied:
Fixed

mik-laj commented on airflow/providers/google/cloud/operators/dataproc.py:
```diff
@@ -155,10 +155,11 @@ class ClusterGenerator:
     """
 
     # pylint: disable=too-many-arguments,too-many-locals
+    @DataprocHook.fallback_to_default_project_id
```
Is it works for operators?

turbaszek replied:
No, it doesn't, so do we want to add such logic? This will require to initialize the hook somewhere so we can access the project_id. I personally don't see a lot of value in it

turbaszek replied:
Or probably we should refactor the operator, so there's no need to provide the project in two places: as argument and in cluster definition

turbaszek committed:
DataprocCreateCluster requires cluster config

In this way users don't have to pass project_id two times (in cluster definition
and as parameter).

turbaszek commented on airflow/providers/google/cloud/operators/dataproc.py:
```diff
@@ -467,20 +454,15 @@ class DataprocCreateClusterOperator(BaseOperator):
     def __init__(  # pylint: disable=too-many-arguments
         self,
         *,
+        cluster_name: str,
```
This backward compatible with 1.10.X:
https://github.com/apache/airflow/blob/827a717fecc675b4d58fa202be003dee8e423632/airflow/contrib/operators/dataproc_operator.py#L192-L193

turbaszek commented on airflow/providers/google/cloud/operators/dataproc.py:
```diff
@@ -506,9 +488,12 @@ def __init__(  # pylint: disable=too-many-arguments
-            cluster = ClusterGenerator(**kwargs).make()
+            if project_id is None:
+                raise AirflowException(
+                    "project_id argument is required when building cluster from keywords parameters"
```
That should not be problem when migrating from 1.10.X becaue it should be in kwargs
https://github.com/apache/airflow/blob/827a717fecc675b4d58fa202be003dee8e423632/airflow/contrib/operators/dataproc_operator.py#L192-L193

turbaszek committed:
Fix Dataproc system test and test helper

olchas commented on airflow/providers/google/cloud/operators/dataproc.py:
```diff
@@ -467,20 +454,15 @@ class DataprocCreateClusterOperator(BaseOperator):
     :type impersonation_chain: Union[str, Sequence[str]]
     """
 
-    template_fields = (
```
Is there a particular reason to move `template_fields` after `__init__` definition?

turbaszek replied:
No particular reason, I will move it to align it with other operators

olchas commented on airflow/providers/google/cloud/operators/dataproc.py:
```diff
@@ -506,9 +488,12 @@ def __init__(  # pylint: disable=too-many-arguments
                 del kwargs['params']
 
             # Create cluster object from kwargs
-            kwargs['region'] = region
```
Is `region` unnecessary for `ClusterGenerator` anymore?

turbaszek replied:
Yes, I will remove it from ClusterGenerator constructor

turbaszek committed:
fixup! Fix Dataproc system test and test helper

turbaszek commented on tests/providers/google/cloud/operators/test_dataproc_system.py:
```diff
@@ -57,12 +52,12 @@ class DataprocExampleDagsTest(GoogleSystemTest):
     @provide_gcp_context(GCP_DATAPROC_KEY)
     def tearDown(self):
-        self.delete_gcs_bucket(BUCKET)
+        # self.delete_gcs_bucket(BUCKET)
```
```suggestion
        self.delete_gcs_bucket(BUCKET)
```

turbaszek committed:
fixup! fixup! Fix Dataproc system test and test helper

turbaszek committed:
Update tests/providers/google/cloud/operators/test_dataproc_system.py
